feat(tip-1067): dynamic base fee (#5153)
tempoxyz/tempo: tips/tip-1067.md

- The block header carries the per-gas base fee (the `baseFeePerGas` field of the Ethereum-derived Tempo header) so that the value computed by this controller is committed to and validated by every node.
- The controller input is the total gas used by all transactions included in a block (the header `gasUsed`), across all lanes (payment and general). The 5,000,000 gas target is compared against this total, so heavy payment-lane usage raises the base fee just as general usage does.
- If an assumption is violated (for example, a header carrying a base fee that does not equal the value this TIP computes), the affected block MUST be rejected as invalid.

---

# Specification

## Constants

| Name | Value | Meaning |
|---|---|---|
| `BASE_FEE_CAP` | `2 × 10^10` attodollars/gas | Maximum base fee; equals TIP-1010's current fixed base fee. |
| `BASE_FEE_FLOOR` | `1 × 10^9` attodollars/gas | Minimum base fee; `BASE_FEE_CAP / 20` (one twentieth of the current fixed cost). |
| `GAS_TARGET` | `5,000,000` gas | Per-block gas usage the controller steers toward. |

> | `BASE_FEE_MAX_CHANGE_DENOMINATOR` | `8` | Bounds the per-block change to at most `1 / 8` (12.5%) of the parent base fee, matching EIP-1559. |

Author: ```suggestion
| `BASE_FEE_MAX_CHANGE_DENOMINATOR` | `8` | Determines the rate of change of the basefee at `1 / 8` (12.5%) of the difference to gas target. |
```

```text
if parent_gas_used == GAS_TARGET:
    base_fee = parent_base_fee

else if parent_gas_used > GAS_TARGET:
    gas_delta   = parent_gas_used - GAS_TARGET
    fee_delta   = max(1, (parent_base_fee * gas_delta)
                          / GAS_TARGET / BASE_FEE_MAX_CHANGE_DENOMINATOR)
    base_fee    = min(parent_base_fee + fee_delta, BASE_FEE_CAP)

else:  # parent_gas_used < GAS_TARGET
    gas_delta   = GAS_TARGET - parent_gas_used
    fee_delta   = (parent_base_fee * gas_delta)
                          / GAS_TARGET / BASE_FEE_MAX_CHANGE_DENOMINATOR
    base_fee    = max(parent_base_fee - fee_delta, BASE_FEE_FLOOR)
```

This is the EIP-1559 update formula with two changes: the gas target is the fixed `GAS_TARGET` constant rather than half of a per-block gas limit, and the result is clamped to `[BASE_FEE_FLOOR, BASE_FEE_CAP]` after the adjustment.

The clamp is applied last, so:

- The base fee can never exceed `BASE_FEE_CAP` (today's fixed cost) regardless of how full blocks are.

> - The base fee can never fall below `BASE_FEE_FLOOR` (one twentieth of today's cost) regardless of how empty blocks are.

Reviewer: Per @malleshpai, 1059-eligible txs should be updated to deduct 60% of the base fee and enforce that the base fee is no less than `BASE_FEE_FLOOR`

Reviewer 2: ACK - we can do that

Author: we are changing this by globally lowering the basefee

---
id: TIP-1067
title: Dynamic Base Fee
description: Replaces Tempo's fixed base fee with an EIP-1559-style controller that lowers the maximum base fee to 1.2e10 attodollars per gas and lowers the base fee toward one twentieth of that cap when the chain is underutilized, with a 10,000,000 gas target.
authors: Dankrad Feist @dankrad

> status: Draft

Reviewer 2: ```suggestion
status: Approved
```

related: TIP-1010, TIP-1000

> protocolVersion: TBD

Reviewer 2: ```suggestion
protocolVersion: T7
```

## Motivation

TIP-1010 sets a single fixed base fee per protocol version and explicitly has no in-protocol mechanism to raise or lower it with utilization; changing it requires a hardfork. That keeps fees predictable but leaves the chain charging the full target price even when it is almost idle, so there is no benefit to users from low utilization.

Tempo is optimized for stablecoin payments and runs fast (~500 ms) blocks with a large block gas limit relative to typical load. For long stretches the chain is expected to be underutilized. During those stretches the marginal cost of including a transaction is minimal, so charging the full target base fee is unnecessary. A bounded dynamic base fee lets the network pass low-utilization savings on to users automatically while preserving a spam floor and never charging more than today's price.

> The design is deliberately a *one-sided discount* relative to the status quo: the cap is below the existing fixed base fee, so this TIP can only ever make fees cheaper than they are now.

Reviewer 3: Can we add some justification to keeping the cap at 1.2. This is almost a 40% discount on all txs from now on with no limit, which increases the DOS concerns of the TIP.

Can't we keep the cap closer to 2, and just make the gas target bigger?
